src/Compilers/Test/Utilities/CSharp/CSharpTestBase.cs

```diff
@@ -1250,18 +1250,50 @@ public static SyntaxTree ParseWithRoundTripCheck(string text, CSharpParseOptions
             return compilation;
         }
 
-        private static CSharpCompilationOptions CheckForTopLevelStatementsAndUnsafe(SyntaxTree[] syntaxTrees)
+        private class DefaultCompilationOptionSelector : SyntaxWalker
         {
-            bool hasTopLevelStatements = syntaxTrees.Any(s => s.GetRoot().ChildNodes().OfType<GlobalStatementSyntax>().Any());
+            private bool _foundTopLevelStatement = false;
+            private bool _foundUnsafe = false;
 
-            var options = hasTopLevelStatements ? TestOptions.ReleaseExe : TestOptions.ReleaseDll;
-            options = options.WithAllowUnsafe(syntaxTrees.Any(containsUnsafe));
+            private DefaultCompilationOptionSelector() : base(SyntaxWalkerDepth.Token)
+            {
+            }
+
+            public static CSharpCompilationOptions GetOptions(SyntaxTree[] syntaxTrees)
+            {
+                var syntaxWalker = new DefaultCompilationOptionSelector();
+
+                foreach (var tree in syntaxTrees)
+                {
+                    syntaxWalker.Visit(tree.GetRoot());
+                    if (syntaxWalker._foundTopLevelStatement && syntaxWalker._foundUnsafe)
+                    {
+                        continue;
+                    }
+                }
 
-            return options;
+                var options = syntaxWalker._foundTopLevelStatement ? TestOptions.ReleaseExe : TestOptions.ReleaseDll;
+                options = options.WithAllowUnsafe(syntaxWalker._foundUnsafe);
 
-            static bool containsUnsafe(SyntaxTree tree)
+                return options;
+            }
+
+            public override void Visit(SyntaxNode node)
             {
-                return tree.GetRoot().DescendantTokens().Any(t => t.IsKind(SyntaxKind.UnsafeKeyword));
+                if (node.Kind() is SyntaxKind.GlobalStatement)
+                {
+                    _foundTopLevelStatement = true;
+                }
+
+                base.Visit(node);
```

Comment:
Consider returning early after `_foundTopLevelStatement` and `_foundUnsafe` are true.